File: src/main/java/boj/problems/step2/No2884.java
package boj.problems.step2;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class No2884 {
    public static final int PRESET_TIME = 45;
    public static final int MIDNIGHT = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter output = new BufferedWriter(new OutputStreamWriter(System.out));

        output.write(solve(input) + "\n");

        input.close();
        output.flush();
        output.close();
    }

    static String solve(BufferedReader input) throws IOException {
        StringTokenizer st = new StringTokenizer(input.readLine());
        
        int hour = Integer.parseInt(st.nextToken());
        int minute = Integer.parseInt(st.nextToken());
        
        if (minute < PRESET_TIME) {
            if (hour == MIDNIGHT) {
                hour = 23;
            } else {
                hour--;
            }
            minute = minute + 60 - PRESET_TIME;

            return hour + " " + minute;
        }

        return hour + " " + (minute - PRESET_TIME);
    }
}

File: src/main/kotlin/boj/problems/step2/No2884.kt
package boj.problems.step2

import java.io.BufferedReader

object No2884 {
    fun solve(input: BufferedReader): String {
        val (n, m) = input.readLine().split(" ").map { it.toInt() }
        val time = Time(n, m).minusPresetTime()
        return time.toString()
    }
}

class Time(private val hour: Int, private val minute: Int) {
    fun minusPresetTime(): Time {
        val totalMinute = hour * MINUTES_IN_HOUR + minute
        val minutesToAdd = if (totalMinute < PRESET_TIME) HOURS_IN_DAY * MINUTES_IN_HOUR else 0
        val newTotalMinute = (totalMinute + minutesToAdd - PRESET_TIME) % (HOURS_IN_DAY * MINUTES_IN_HOUR)
        val newHour = newTotalMinute / MINUTES_IN_HOUR
        val newMinute = newTotalMinute % MINUTES_IN_HOUR
        return Time(newHour, newMinute)
    }

    override fun toString(): String {
        return "$hour $minute"
    }

    companion object {
        private const val PRESET_TIME = 45
        private const val HOURS_IN_DAY = 24
        private const val MINUTES_IN_HOUR = 60
    }
}
